Allow user to set up a lambda to initialize new connections (#705)
awslabs/aws-advanced-jdbc-wrapper: docs/using-the-jdbc-driver/UsingTheJdbcDriver.md

        CustomConnectionPluginFactory.class));
```

> ### Connection Initialization

Reviewer: ## Executing Custom Code When Initializing a Connection

> In some cases it's necessary to configure a connection before a user application can use it. Some target drivers provides such functionality and allow to specify a configuration parameter with SQL statements that are executed when connection is established. However, not all drivers supports such functionality. Also, in some cases, additional conditions should be checked in order to identify what initialization is required for a particular connection.

Reviewer: In some particular cases, users may need to define a specific configuration for a connection that has just been opened by the driver before an application can use it.

Some...

> In some cases it's necessary to configure a connection before a user application can use it. Some target drivers provides such functionality and allow to specify a configuration parameter with SQL statements that are executed when connection is established. However, not all drivers supports such functionality. Also, in some cases, additional conditions should be checked in order to identify what initialization is required for a particular connection.

Reviewer 2: ```suggestion
Users may need to define a specific configuration for a connection that has just been opened by the driver before an application can use it. Not all target drivers provide this functionality, but some allow specifying a configuration parameter with SQL statements that are executed when a connection is established. Some cases may also require that additional conditions are checked in order to identify what initialization configuration is required for a particular connection.
```
?

> Users may need to define a specific configuration for a connection that has just been opened by the driver before an application can use it. Not all target drivers provide this functionality, but some allow specifying a configuration parameter with SQL statements that are executed when a connection is established. Some cases may also require that additional conditions are checked in order to identify what initialization configuration is required for a particular connection.

Reviewer 3: Thoughts on rewording to the following?
```suggestion
In some use cases you may need to define a specific configuration for a new driver connection before your application can use it. For instance:
- you might need to run some initial SQL queries when a connection is established, or;
- you might need to check for some additional conditions to determine the initialization configuration required for a particular connection.
```

Reviewer 3: Does it make sense to link to the issue here for a concrete use case?

AWS JDBC Driver allows to specify a special function that can initialize a connection. It can be done with `ConnectionProviderManager.setConnectionInitFunc` method. `resetConnectionInitFunc` method is also available. 

> The initialization function is called for all connections, including pre-opened connections provided by internal connection pool (see [Using Read Write Splitting Plugin Internal Connection Pooling](./using-plugins/UsingTheReadWriteSplittingPlugin.md#internal-connection-pooling)), and, thus, helping a user application to clean up connection session "contaminated" by previous use.

Author: A better wording ideas?

> The initialization function is called for all connections, including pre-opened connections provided by internal connection pool (see [Using Read Write Splitting Plugin Internal Connection Pooling](./using-plugins/UsingTheReadWriteSplittingPlugin.md#internal-connection-pooling)), and, thus, helping a user application to clean up connection session "contaminated" by previous use.

Reviewer 2: ```suggestion
The initialization function is called for all connections, including pre-opened connections provided by internal connection pools (see [Using Read Write Splitting Plugin and Internal Connection Pooling](./using-plugins/UsingTheReadWriteSplittingPlugin.md#internal-connection-pooling)). This helps user applications clean up connection sessions that have been altered by previous operation, as returning a connection to a pool will reset the state and retrieving it will call the initialization function again.
```
Does this make sense? I'm not sure if this was the original intent of this statement

> The initialization function is called for all connections, including pre-opened connections provided by internal connection pools (see [Using Read Write Splitting Plugin and Internal Connection Pooling](./using-plugins/UsingTheReadWriteSplittingPlugin.md#internal-connection-pooling)). This helps user applications clean up connection sessions that have been altered by previous operation, as returning a connection to a pool will reset the state and retrieving it will call the initialization function again.

Reviewer 3: ```suggestion
The initialization function is called for all connections, including connections opened by the internal connection pools (see [Using Read Write Splitting Plugin and Internal Connection Pooling](./using-plugins/UsingTheReadWriteSplittingPlugin.md#internal-connection-pooling)). This helps user applications clean up connection sessions that have been altered by previous operations, as returning a connection to a pool will reset the state and retrieving it will call the initialization function again.
```

Reviewer 4: We're saying that returning a connection to the pool resets its session state, is this accurate though? If I understand this PR correctly, the only extra logic is performed when retrieving a connection, not when returning it, and whether the session state is fully reset or not depends on what the user defines in their ConnectionInitFunc object.

Author: Hikari CP does reset some connection/session settings while returning connection back to the pool. Hikari tracks just a limited number of session settings like read-only, transaction isolation, etc. It doesn't track temporary tables, language settings and other that can be essential for next user transactions.  This function is about setting them while getting connection from the pool, or while opening a brand new connection.

> > :warning: Executing CPU and network intensive code in the initialization function may have a significant impact in the wrapper performance overall.

Reviewer 3: ```suggestion
> [!WARNING]\
> Executing CPU and network intensive code in the initialization function may significantly impact the AWS JDBC Driver's overall performance.
```
